Feature branch

cristianCeamatu opened the pull request:
This is a private events app following the project requirements from TheOdinProject website. The project enforced our skills in using Model associations and create from scratch Models, Controllers, Views, and Migrations.

**App functionality**
- Users can be created, they can sign in and sign out
- Users can create and attend events (they cannot attend events they are already attending)
- Events are categorized in past and upcoming events based on their date

cristianCeamatu committed:
Add user controller, model and views

cristianCeamatu committed:
Added basic login functionality for the user

uimarshall committed:
Add event model and controllers

uimarshall committed:
Add fields to index and show page

uimarshall committed:
Add model for event_attendee and associations

cristianCeamatu committed:
Readme updates and User-Events association

cristianCeamatu committed:
Add
Past events and upcoming events functionality
Detailed view for events#index, user#show, and events#show

cristianCeamatu committed:
Update readme.details

uimarshall committed:
Add bootstrap and include navbar

cristianCeamatu committed:
Rspec initial setup

uimarshall committed:
Add styles to index and create event page

uimarshall committed:
Styled user created and show page

cristianCeamatu committed:
Merge branch 'feature_branch' of github.com:cristianCeamatu/microverse-rails-private-events into rspec_branch

cristianCeamatu committed:
Small touches on the UI

cristianCeamatu committed:
Rspec unit and integrations tests

cristianCeamatu committed:
Add
Final touches on rspec
Heroku changes

cristianCeamatu committed:
Heroku changes

cristianCeamatu committed:
Changed adaptor for production

cristianCeamatu committed:
Heroku changes

cristianCeamatu committed:
Heroku changes

cristianCeamatu committed:
Live demo link

addod19 commented on app/models/event.rb:
```diff
@@ -0,0 +1,12 @@
+  has_many :attendees, through: :event_attendees, source: :event_attendee
+  belongs_to :creator, class_name: 'User'
+
+  scope :upcoming_events, -> { where('date > ?', Date.today) }
```
Kindly replace `'date > ?'` with `'date >= ?'`

addod19 reviewed (changes requested):
## Status: Require changes :recycle: :traffic_light: 

Hi, @cristianCeamatu and @uimarshall, great job with this project :+1: README is quite detailed, PR is professional, has a title and a short description:smile: KIndly address the comments below so we get our project approved :+1: 
- Let us add the migration command before we run `rails s`. Otherwise, we will run into the `Pending migration error`
- We are currently missing the `foreign keys` in our migration file. Let us check our migration files so we can fix them. Kindly check out this [link](
https://api.rubyonrails.org/v5.1.7/classes/ActiveRecord/ConnectionAdapters/TableDefinition.html#method-i-references) or [this](
https://api.rubyonrails.org/v5.1.7/classes/ActiveRecord/ConnectionAdapters/TableDefinition.html#method-i-belongs_to) in adding foreign keys
- `Events Controller`, our `Index` action is missing the following;
  - [x] `@past_events = Event.previous_events`
  - [x] `@upcoming_events = Event.upcoming_events`
- `Users Controller`, our `Show` action is missing the following:
  - [x] List events created by a user `@created_events = current_user.events`
  - [x] List event previously attended by a user ` @past_events = current_user.attended_events.previous `
  - [x] List upcoming events previously attended by a user using event model scope `
@upcoming_events = current_user.attended_events.upcoming`
- `Test`, currently, the `rspec` command as outlined in the README file is not working. Kindly run `rails db:migrate RAILS_ENV=test` before running the `rspec`
We currently have some test failing, kindly check the screenshot below to understand the issue
![ftest](https://user-images.githubusercontent.com/46388113/92968671-15a47980-f46b-11ea-8d38-5af40b403615.png)
- Events created on the same day is not being recorded, kindly fix that in the `event scope model`

Kindly address the issue above, once these are fixed, kindly submit another code review :slightly_smiling_face: 
_Happy coding_ :man_technologist: 
[Daniel Larbi Addo](linkedin.com/in/daniel-larbi-addo/)

cristianCeamatu committed:
Fix TSE requests
Add migrations commands in the README.md
Modified EventAttendees table to have foreign keys instead of regular integers columns
Updated Events and Users controllers to remove logic data from the views
Added Tests for the EventAttendees module and associations
Bug that was not showing events from today on the index page

cristianCeamatu commented:
@addod19  Thank you for a very constructive code review.
Everything updated as per requests, have a great day!

maelfosso commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,74 @@
+        <% end %>
+      <% else %>
+        <p>No upcoming events by <%= @user.name %></p>
+      <% end %>
```
Logic not allowed in views. Kindly implement them in views helpers.

maelfosso commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,74 @@
+          <% end %>
+        <% else %>
+          <p>No past events by <%= @user.name %></p>
+        <% end %>
```
Logic not allowed in views. Kindly implement them in views helpers.

maelfosso commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,74 @@
+          <p>No past events</p>
+        <% end %>
+      </div>
+    </section>
```
Logic not allowed in views. Kindly implement them in views helpers.

maelfosso commented on app/views/layouts/application.html.erb:
```diff
@@ -1,15 +1,42 @@
+          <li><%= link_to 'Private Events', root_path, class: 'text-white font-weight-bold' %></li>
+        </ul>
+        <form class="form-inline my-2 my-lg-0">
+          <% if !session[:user_id].nil? %>
```
Logic not allowed in views. Kindly implement them in views helpers.

maelfosso replied:
It's preferable to create a helper called `current_user` instead of using `session[:user_id]`

maelfosso reviewed (changes requested):
# Changes required :red_circle: 

Hello,

Thank you for applying the suggested changes.
However, some requirements are missings

1- You've forgotten to add indexes in `Attendance` table

2- Kindly remove ogic from views. Read the comment leff below

3- For the unit tests, I suggest you to use the **shoulda-matcher** gem which is really helpful

Kindly make the suggested changes and submit a new code review request
Thanks

Mael FOSSO

cristianCeamatu committed:
Add
Indexes to foreign keys
Removed logic from views
current_user helper and replaced the session[:user_id] condition in the views

maelfosso commented on app/views/layouts/application.html.erb:
```diff
@@ -1,15 +1,42 @@
+          <li><%= link_to 'Private Events', root_path, class: 'text-white font-weight-bold' %></li>
+        </ul>
+        <form class="form-inline my-2 my-lg-0">
+          <% if current_user %>
+            <%= link_to "Signed In as " + @user.name, user_path(@user),class: 'text-white ml-3'  %>
+            <%= link_to 'All events', events_path, class: 'btn btn-light my-2 my-sm-0 text-info ml-2' %>
+            <%= link_to 'Create event', new_event_path, class: 'btn btn-light my-2 my-sm-0 text-info ml-3' %>
+            <%= link_to "Sign out", sign_out_path, method: :delete, class: 'btn btn-light my-2 my-sm-0 text-info ml-2' %>
+          <% else %>
+            <%= link_to 'All events', events_path, class: 'btn btn-light my-2 my-sm-0 text-info ml-2' %>
```
There are still logic in your code. please, move them into views helpers

cristianCeamatu replied:
All the tutorials and reading materials we followed up used if/else statements in their views, also the #any? and #each methods. This is the first time we see that you cannot use IF/ELSE statements in your view. Below is a screenshot of an official RAILCASTA video where they are doing exactly what we are doing in our project.
![20200915_083411](https://user-images.githubusercontent.com/19146884/93174506-be3a2e00-f736-11ea-8400-4a3d436e7f20.jpg)

maelfosso commented on app/views/events/index.html.erb:
```diff
@@ -0,0 +1,59 @@
+          <% end %>
+        </div>
+
+        <% if current_user && !@user.attended_events.map { |el| el.id }.include?(event.id) %>
```
That is first logic in views and it must be moved into view helpers. Secondly, that is a request into views which is not allowed.
Move that into  `Event` model and call it after inside your view helper

maelfosso commented on spec/models/event_spec.rb:
```diff
@@ -0,0 +1,85 @@
+      expect(@event1).to_not be_valid
+    end
+
+    # it 'should be valid if event:description is nil' do
+    #   @event.description = nil
+    #   @event.save
+    #   expect(@event).to_not be_valid
+    # end
+
+    # it 'should be valid if event:description is below min length' do
+    #   @event.description = 'xxx'
+    #   @event.save
+    #   expect(@event).to_not be_valid
+    # end
+
+    # it 'should be valid if event:place is nil' do
+    #   @event.place = nil
+    #   @event.save
+    #   expect(@event).to_not be_valid
+    # end
+
+    # it 'should be valid if event:place is below min length' do
+    #   @event.place = 'xxx'
+    #   @event.save
+    #   expect(@event).to_not be_valid
+    # end
+
+    # it 'should be valid if event:date is nil' do
+    #   @event.date = nil
+    #   @event.save
+    #   expect(@event).to_not be_valid
+    # end
```
Remove the comments

maelfosso reviewed (changes requested):
# Changes required :red_circle: 

Hello, 

Thank you for applying the suggested changes.
However, you missed some.
Kindly read the comment left below, make the suggested changes and submit a new code review request

Thanks

Mael FOSSO

cristianCeamatu committed:
TSE requests
Remove comments from the spec file
Remove logic from views

cristianCeamatu committed:
Rubocop errors

wathigo reviewed (approved):
### Approved :heavy_check_mark: 
Great job on this project.

**You may move on.**

**Good Luck :four_leaf_clover: 
Simon Wathigo.**
